pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,287 @@
+	CoreClient v1core.CoreV1Interface
+
+	// AuthenticationClient is used to check API tokens to make sure they are valid before
+	// building a controller client from them
+	AuthenticationClient v1authentication.AuthenticationV1Interface
+
+	// Namespace is the namespace used to host the service accounts that will back the
+	// controllers.  It must be highly privileged namespace which normal users cannot inspect.
+	Namespace string
+
+	tokenSourceMap map[string]oauth2.TokenSource
+
+	mutex sync.RWMutex
+
+	clock clock.Clock
+}
+
+func NewDynamicClientBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		tokenSourceMap:       map[string]oauth2.TokenSource{},
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(saName string) (*restclient.Config, error) {
+	_, err := t.getOrCreateServiceAccount(saName)
+	if err != nil {
+		return nil, err
+	}
+
+	configCopy := constructClient(t.Namespace, saName, t.ClientConfig)
+
+	ts, ok := t.get(saName)
+	if ok {
+		configCopy.TokenSource = ts
+	} else {
+		configCopy.TokenSource = &tokenSourceImpl{
+			clientConfig:       t.ClientConfig,
+			namespace:          t.Namespace,
+			serviceAccountName: saName,
+			coreClient:         t.CoreClient,
+			authClient:         t.AuthenticationClient,
+			clock:              clock.RealClock{},
+		}
+		t.set(saName, configCopy.TokenSource)
+	}
+
+	return &configCopy, nil
+}
+
+func (t *DynamicControllerClientBuilder) ConfigOrDie(name string) *restclient.Config {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return clientConfig
+}
+
+func (t *DynamicControllerClientBuilder) Client(name string) (clientset.Interface, error) {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		return nil, err
+	}
+	return clientset.NewForConfig(clientConfig)
+}
+
+func (t *DynamicControllerClientBuilder) ClientOrDie(name string) clientset.Interface {
+	client, err := t.Client(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return client
+}
+
+func (t *DynamicControllerClientBuilder) getOrCreateServiceAccount(name string) (*v1.ServiceAccount, error) {
+	sa, err := t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	if err == nil {
+		return sa, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); err != nil {
+		t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}})
```

Comment:
Check the error?

Reply:
Maybe a warning log is enough? I think we should continue here even error happens.